certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataConverter.java

```diff
@@ -0,0 +1,114 @@
+package se.inera.intyg.certificateservice.application.certificate.service;
+
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+import se.inera.intyg.certificateservice.application.certificate.dto.CertificateDataElement;
+import se.inera.intyg.certificateservice.domain.certificate.model.ElementData;
+import se.inera.intyg.certificateservice.domain.certificate.model.ElementValue;
+import se.inera.intyg.certificateservice.domain.certificatemodel.model.CertificateModel;
+import se.inera.intyg.certificateservice.domain.certificatemodel.model.ElementId;
+import se.inera.intyg.certificateservice.domain.certificatemodel.model.ElementSpecification;
+
+@Component
+@RequiredArgsConstructor
+public class CertificateDataConverter {
+
+  private final CertificateDataValueConverter certificateDataValueConverter;
+  private final CertificateDataConfigConverter certificateDataConfigConverter;
+  private final CertificateDataValidationConverter certificateDataValidationConverter;
+
+  public Map<String, CertificateDataElement> convert(CertificateModel certificateModel,
```

Comment:
Good job! It's a bit complicated code, but it does the work and you have build a great unit test suite to verify that. 

What I would do is try to flatten and convert at the same time. I think that would simplify the code and make it more readable. It won't be too simple though, because we need to use recursive functions.

I don't expect you to change it and we can go ahead with this. But if you like to take a look at it, here is an article that flattens a hierarchical list structure: https://stackoverflow.com/questions/52670243/recursively-flatten-a-list-with-streams

Reply:
I made a try at flattening out the code, and I think it improved the readability of it. Lets discuss improvements further on monday!